ctapipe/reco/sklearn.py

```diff
@@ -507,6 +512,353 @@ def predict_table(self, key, table: Table) -> Table:
+        """
+        feature_table = table[self.features]
+        valid = check_valid_rows(feature_table, log=self.log)
+        X = table_to_float(feature_table[valid])
+        return X, valid
+
+    def _table_to_y(self, table, mask=None):
+        """
+        Extract target values as numpy array from input table
+        """
+        # make sure we use the unit that was used during training
+        if self.norm_unit is not None:
+            norm_y = table[mask][self.norm_target].quantity.to_value(self.norm_unit)
+        else:
+            norm_y = np.array(table[self.norm_target][mask])
+
+        if self.norm_log_target:
+            if np.any(norm_y <= 0):
+                raise ValueError("norm_y contains negative values, cannot apply log")
+            norm_y = np.log(norm_y)
+
+        sign_y = np.array(table[self.sign_target][mask])
+        return norm_y, sign_y
+
+    def fit(self, key, table):
+        """
+        Create and fit new models for ``key`` using the data in ``table``.
+        """
+        self._norm_models[key], self._sign_models[key] = self._new_models()
+
+        X, valid = self._table_to_X(table)
+        self.norm_unit = table[self.norm_target].unit
+        norm_y, sign_y = self._table_to_y(table, mask=valid)
+        self._norm_models[key].fit(X, norm_y)
+        self._sign_models[key].fit(X, sign_y)
+
+    def write(self, path, overwrite=False):
+        path = pathlib.Path(path)
+
+        if path.exists() and not overwrite:
+            raise IOError(f"Path {path} exists and overwrite=False")
+
+        with path.open("wb") as f:
+            Provenance().add_output_file(path, role="ml-models")
+            joblib.dump(self, f, compress=True)
+
+    @classmethod
+    def read(cls, path, **kwargs):
+        with open(path, "rb") as f:
+            instance = joblib.load(f)
+
+        for attr, value in kwargs.items():
+            setattr(instance, attr, value)
+
+        if not isinstance(instance, cls):
+            raise TypeError(
+                f"{path} did not contain an instance of {cls}, got {instance}"
+            )
+
+        Provenance().add_input_file(path, role="ml-models")
+        return instance
+
+    @lazyproperty
+    def instrument_table(self):
+        return self.subarray.to_table("joined")
+
+    def _predict(self, key, table):
+        if key not in self._norm_models:
+            raise KeyError(
+                f"No regressor available for key {key},"
+                f" available regressors: {self._norm_models.keys()}"
+            )
+        if key not in self._sign_models:
+            raise KeyError(
+                f"No classifier available for key {key},"
+                f" available classifiers: {self._sign_models.keys()}"
+            )
+        X, valid = self._table_to_X(table)
+        norm_prediction = np.full(len(table), np.nan)
+        sign_prediction = np.full(len(table), self.sign_invalid_class, dtype=np.int8)
```

Comment:
Ah! forget comment above.  But I think the option is not necessary, filling 0 for unknown sign shouldn't something we want to change, right?

Reply:
You mean your comment regarding changing `sign_invalid_class` to `nan`, correct? So setting it to `0` permanently should be fine, right?

Reply:
Yes, but I'd even prefer the solution to just store a single disp value with sign included (and setting that to nan for `not is_valid)`